plugins/pulp_docker/plugins/importers/upload.py

```diff
@@ -169,6 +179,47 @@ def get_models(self, metadata, mask_id=''):
         return images
 
 
+class ImportManifestList(PluginStep):
+    """
+    Add an uploaded ManifestList to a repository.
+    """
+
+    def process_main(self):
+        """
+        Validate the uploaded manifest list json, then import content unit into repository.
+        """
+        with open(self.parent.file_path, 'r') as uploaded_file:
+            manifest_list = uploaded_file.read()
+        models.ManifestList.validate_json(manifest_list)
+        digest = models.Manifest.calculate_digest(manifest_list)
+        manifest_list_instance = models.ManifestList.from_json(manifest_list, digest)
+        transfer_repo = self.get_repo()
+
+        #  Ensure that all referenced manifests are already in repository.
+        manifest_digests = set()
+        for manifest in manifest_list_instance.manifests:
+            manifest_digests.add(manifest)
+        qs = models.Manifest.objects.filter(
+            digest__in=sorted(manifest_digests)).only('id', 'digest')
+        referenced_manifests = {manifest['digest']: manifest['id'] for manifest in qs}
```

Comment:
idea - to not perform further query and loop, at this point, we can make set from referenced_manifests, make a difference from set of manifest_digests, and raise the missing digests which we see in the difference of 2 sets

Reply:
At the time that you made this comment, my goal was to fail early with any missing Manifest List., and this is a place we can fail earlier. However, after considering this, I think the better thing to do is to detect all failures. For the cost of only 1 extra call to the db per failed Manifest List upload, an upload failure can return a complete list of Manifests that are missing from the repository. In the event that this error is raised, I'm betting that the user would rather know all the things that must be done than know only 1 missing thing a fraction of a second earlier.